Source/Core/Common/File.cpp

```diff
@@ -0,0 +1,335 @@
+                             std::char_traits<char>::to_int_type(*gptr());
+}
+
+StandardDirectoryIterator::StandardDirectoryIterator(const std::string& path)
+{
+#ifdef _WIN32
+  m_handle = FindFirstFile(UTF8ToTStr(path + "\\*").c_str(), &m_find_data);
+#else
+  m_dir = opendir(directory.c_str());
+#endif
+}
+
+StandardDirectoryIterator::~StandardDirectoryIterator()
+{
+#ifdef _WIN32
+  FindClose(m_handle);
+#else
+  if (m_dir)
+    closedir(m_dir);
+#endif
+}
+
+std::string StandardDirectoryIterator::NextChild()
+{
+  std::string name;
+#ifdef _WIN32
+  if (m_handle != INVALID_HANDLE_VALUE)
+  {
+    if (m_first_child)
+    {
+      name = TStrToUTF8(m_find_data.cFileName);
+m_first_child = false;
+    }
+    else if (FindNextFile(m_handle, &m_find_data))
+    {
+      name = TStrToUTF8(m_find_data.cFileName);
+    }
+  }
+#else
+if (m_dir)
+{
+  dirent* result = readdir(m_dir);
+  if (result)
+    name = result->d_name;
+}
+#endif
+
+if (name == "." || name == "..")
+return NextChild();
+
+return name;
+}
+
+Path::const_iterator::const_iterator(const Path& path)
+  : m_base_path(path.m_path), m_child_path(std::make_shared<Path>(path))
+{
+  m_iterator = std::make_shared<StandardDirectoryIterator>(path.m_path);
+}
+
+Path::const_iterator& Path::const_iterator::operator++()
+{
+  std::string child_name = m_iterator->NextChild();
+  if (child_name.empty())
+    m_child_path = std::make_unique<Path>();
+  else
+    m_child_path->m_path = m_base_path + "/" + child_name;
+
+  return *this;
+}
+
+Path Path::operator+(const std::string& str) const
+{
+  return Path(m_path + str);
+}
+
+Path& Path::operator+=(const std::string& str)
+{
+  m_path += str;
+  return *this;
+}
+
+std::string Path::GetName() const
+{
+  std::string name;
+  SplitPath(m_path, nullptr, &name);
+  return name;
+}
+
+bool Path::Exists() const
+{
+  struct stat file_info;
+  std::string path = StripTailDirSlashes(m_path);
+
+#ifdef _WIN32
+  int result = _tstat64(UTF8ToTStr(path).c_str(), &file_info);
+#else
+  int result = stat(copy.c_str(), &file_info);
+#endif
+
+  return result == 0;
+}
+
+bool Path::IsDirectory() const
+{
+  struct stat file_info;
+  std::string path = StripTailDirSlashes(m_path);
+
+#ifdef _WIN32
+  int result = _tstat64(UTF8ToTStr(path).c_str(), &file_info);
+#else
+  int result = stat(copy.c_str(), &file_info);
+#endif
+
+  return result == 0 && (file_info.st_mode & S_IFMT) == S_IFDIR;
+}
+
+u64 Path::GetFileSize() const
+{
+  struct stat file_info;
+#ifdef _WIN32
+  int result = _tstat64(UTF8ToTStr(m_path).c_str(), &file_info);
+#else
+  int result = stat(filename.c_str(), &buf);
+#endif
+
+  if (result != 0)
+  {
+    WARN_LOG(COMMON, "GetSize: failed: No such file");
+    return 0;
+  }
+  else if ((file_info.st_mode & S_IFMT) == S_IFDIR)
+  {
+    WARN_LOG(COMMON, "GetSize: failed: is a directory");
+    return 0;
+  }
+
+  return file_info.st_size;
+}
+
+std::unique_ptr<ReadOnlyFile> Path::OpenFile(bool binary) const
```

Comment:
I can't return it by value because it can be either an IOFile or an AndroidAsset (which hasn't been implemented yet, hence the WIP tag).